Support level tags for external contributions bug fix

GuyAfik opened the pull request:
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-6847)

## Description
* Fixed an issue where contributions with new packs didn't add support labels at all.
For more info refer to issues comment here: https://jira-hq.paloaltonetworks.local/browse/CIAC-6847?focusedCommentId=6979394&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-6979394
* Added to the **Checkout** context manager the ability to checkout to forked branches as well.


## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

GuyAfik commented:
please see the following external PR which fixes the problem: https://github.com/demisto/content/pull/27366 and its workflow: https://github.com/demisto/content/actions/runs/5235827205/jobs/9452945573?pr=27366

ShacharKidor commented:
@daryakoval  - Would appreciate your review here since we changed a code that is used for the auto bump mechanism.

ShacharKidor commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,53 @@ class Checkout:  # pragma: no cover
-        self._branch_to_checkout = branch_to_checkout
+
+        if fork_owner:
+            forked_remote_name = f'{fork_owner}_{branch_to_checkout}_remote'
```
what is this name? just a name we use internally in the code or it is based on something related to contrib PRs?

GuyAfik replied:
its the name of that remote that will be created in the workflow to the given repository of the forked contributor branch.

You can choose any name you want, in our case I thought its best to include the `fork_owner` + `branch_to_checkout` and maybe ill add also the name of the repo that we checkout to it.

ShacharKidor reviewed (changes requested):
Good work!

tkatzir commented on Utils/github_workflow_scripts/handle_external_pr.py:
```diff
@@ -64,7 +67,26 @@ def determine_reviewer(potential_reviewers: List[str], repo: Repository) -> str:
 
 
-def get_packs_support_level_label(file_paths: List[str]) -> str:
+def get_packs_support_levels(pack_dirs: Set[str]) -> Set[str]:
```
`Set`s don't retain order.
Wouldn't you rather use `list`s, or return a `dict`, so you could match `pack_dir` to `pack_support_level`?

GuyAfik replied:
i don't really care about order here, its enough to get the support levels as a set and print out packs that we couldn't get their support labels.

tkatzir commented on Utils/github_workflow_scripts/handle_external_pr.py:
```diff
@@ -79,32 +101,39 @@ def get_packs_support_level_label(file_paths: List[str]) -> str:
         highest support level of the packs that were changed, empty string in case no packs were changed.
     """
-    changed_pack_dirs = set()
+    pack_dirs_to_check_support_labels = set()
```
Rename `pack_dirs_to_check_support_labels` to either one of:

- `pack_dirs_to_check_support_levels`
- `pack_dirs_to_check_support_levels_labels`

tkatzir commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,53 @@ class Checkout:  # pragma: no cover
+
+        if fork_owner:
+            forked_remote_name = f'{fork_owner}_{branch_to_checkout}_remote'
+            url = f"https://github.com/{fork_owner}/content"
```
`"content"` should be a param too, with a default value.

tkatzir reviewed (commented):
Looks good!
Some notes and comments.

daryakoval commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,55 @@ class Checkout:  # pragma: no cover
     def __enter__(self):
         """Checks out the given branch"""
-        self.repo.git.checkout(self._branch_to_checkout)
-        return self
```
Return self for being able to do: 
``` with Checkout() as c:```

GuyAfik replied:
True, I guess i added self by mistake, however i do not see any harm in adding it, you can still use it `with Checkout()`.

 in addition you deleted the 
`self.repo.git.checkout(self._branch_to_checkout)` by mistake in the suggestion which i believe you didn't mean it.

daryakoval replied:
I think the `return self` is for `with Checkout() as c:` "as c" part.

GuyAfik replied:
i understood what you meant

daryakoval commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,55 @@ class Checkout:  # pragma: no cover
+            try:
+                self.repo.create_remote(name=forked_remote_name, url=url)
+            except Exception as error:
+                print(f'could not create remote from {url}, {error=}')
```
Why not fail here?

GuyAfik replied:
failing in line 124 in case the `GITHUB_EVENT_PATH` environment variable does not exist.

it is possible that the forked repository name is not actually `content` so this is the case being handled in the exception taking advantage of the `GITHUB_EVENT_PATH` to get details on the forked repo.

daryakoval commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,55 @@ class Checkout:  # pragma: no cover
+                    forked_repo_name = payload.get("pull_request", {}).get("head", {}).get("repo", {}).get("full_name")
+                    self.repo.create_remote(name=forked_remote_name, url=f"https://github.com/{forked_repo_name}")
+                else:
+                    raise
```
What is the error that will be printed in this case?

GuyAfik replied:
the exception from line 111.

daryakoval reviewed (commented):
Hi, please see my comments. 
Before merging please test the flow using the different options you added.

daryakoval commented on Utils/github_workflow_scripts/utils.py:
```diff
@@ -92,25 +92,55 @@ class Checkout:  # pragma: no cover
-        self.repo.git.checkout(self._branch_to_checkout)
-        return self
+        self.repo.git.checkout(self.branch_to_checkout)
+        print(f'Checked out to branch {self.branch_to_checkout}')
```
```suggestion
        print(f'Checked out to branch {self.branch_to_checkout}')
        return self
```

GuyAfik commented:
@daryakoval i will check it in depth before merge, at the moment this task is not in high priority, **will not be merged before in-depth check-ups**

tkatzir reviewed (approved):
Looks great!

GuyAfik commented:
here is a full flow showing its working: https://github.com/demisto/content/pull/27473

added a new pack with `xsoar` supported label and changed an existing partner pack to be `community`, the workflow was able to checkout to the forked branch and find both of them, ending in adding the `Xsoar Support Level`.

@ShacharKidor fyi
